runtime/src/accounts_db.rs

```diff
@@ -969,28 +973,49 @@ impl AccountsDB {
         datapoint_info!("accounts_db-stores", ("total_count", total_count, i64));
     }
 
-    pub fn verify_hash_internal_state(&self, slot: Slot, ancestors: &HashMap<Slot, usize>) -> bool {
-        let mut hash_state = BankHash::default();
-        let hashes: Vec<_> = self.scan_accounts(
+    pub fn verify_bank_hash(
+        &self,
+        slot: Slot,
+        ancestors: &HashMap<Slot, usize>,
+    ) -> Result<(), BankHashVerificatonError> {
+        use BankHashVerificatonError::*;
+
+        let (hashes, mismatch_found) = self.scan_accounts(
             ancestors,
-            |collector: &mut Vec<BankHash>, option: Option<(&Pubkey, Account, Slot)>| {
+            |(collector, mismatch_found): &mut (Vec<BankHash>, bool),
+             option: Option<(&Pubkey, Account, Slot)>| {
                 if let Some((pubkey, account, slot)) = option {
                     if !sysvar::check_id(&account.owner) {
-                        let hash = BankHash::from_hash(&Self::hash_account(slot, &account, pubkey));
+                        let hash = Self::hash_account(slot, &account, pubkey);
+                        if hash != account.hash {
+                            *mismatch_found = true;
```

Comment:
[Originally assigned `false`](https://github.com/solana-labs/solana/pull/7427/files#diff-2099c5256db4eb5975c8834af38f6456R937), but this didn't work because the `Default` of `boolean` is `false`... So I reverted the boolean.